segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MetadataTransaction.java

```diff
@@ -0,0 +1,229 @@
+        store.create(this, metadata);
+    }
+
+    /**
+     * Marks given record as pinned.
+     *
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void markPinned(StorageMetadata metadata) throws StorageMetadataException {
+        store.markPinned(this, metadata);
+    }
+
+    /**
+     * Deletes a metadata record given the key.
+     *
+     * @param key key to use to retrieve metadata.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    public void delete(String key) throws StorageMetadataException {
+        store.delete(this, key);
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void commit() throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this);
+        isCommitted = true;
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @param lazyWrite true if data can be written lazily.
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void commit(boolean lazyWrite) throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this, lazyWrite);
+        isCommitted = true;
+    }
+
+    /**
+     * Commits the transaction.
+     *
+     * @param lazyWrite      true if data can be written lazily.
+     * @param skipStoreCheck true if data is not to be reloaded from store.
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void commit(boolean lazyWrite, boolean skipStoreCheck) throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        store.commit(this, lazyWrite, skipStoreCheck);
+        isCommitted = true;
+    }
+
+    /**
+     * Aborts the transaction.
+     *
+     * @throws StorageMetadataException If transaction could not be commited.
+     */
+    public void abort() throws StorageMetadataException {
+        Preconditions.checkState(!isCommitted, "Transaction is already committed");
+        Preconditions.checkState(!isAborted, "Transaction is already aborted");
+        isAborted = true;
+        store.abort(this);
+    }
+
+    /**
+     * {@link AutoCloseable#close()} implementation.
+     *
+     * @throws Exception In case of any errors.
+     */
+    public void close() throws Exception {
+        if (!isCommitted || isAborted) {
+            store.abort(this);
```

Comment:
shall we log that we are automatically trying to abort the transaction ?
what happens in case of an error ?

what happens if you are calling this method twice ? will the *store* throw an error ?

Reply:
the abort implementation in BaseMetadataStore is empty. There is no action required. TableBasedMetadataStore or InMemoryMetadataStore do not override it. If some derived class overrides it then it may throw an error. In this case I would expect `store.abort();` to be idempotent.